### PR TITLE
Improve doc comment

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
I took another pass at updating and expanding the package doc. Comments appreciated @kortschak  @vladimir-ch. My goal was to highlight the features of the package and help a new user get started quickly. In the example codes, I wanted to highlight a variety of the features provided by the package. In addition to word choice issues, if there are additional major sections that should be discussed, please let me know.

File: mat64/doc.go [hunk 1]
```diff
@@ -2,38 +2,119 @@
...
+//  // Allocate a zeroed array of size 3×5
+//  zero := mat64.NewDense(3, 5, nil)
+// If a backing data slice is provided, the matrix will have those elements.
+// Matrices are all stored in row-major format.
+//  // Generate a 6×6 matrix of random values.
+//  data := make([]float64, 36)
+//  for i := range data {
+//		data[i] = rand.NormFloat64()
+//  }
+//  a := mat64.NewDense(6,6, data)
```
Comment (vladimir-ch) [inline mat64/doc.go]: space

File: mat64/doc.go [hunk 2]
```diff
@@ -2,38 +2,119 @@
...
+// will be modified to have the correct size, allocating data if necessary.
+//  var c mat64.Dense // construct a new zero-sized matrix
+//  c.Mul(a, a)       // c is automatically adjusted to be 6×6
+//
+// The Matrix Interface(s)
+//
+// The Matrix interface is the common link between the concrete types. The Matrix
+// interface is defined by three functions: Dims, which returns the dimensions
+// of the Matrix, At, which returns the element in the specified location, and
+// T, which will be discussed later. All of the concrete types can perform these
```
Comment (vladimir-ch) [inline mat64/doc.go]: I would add what T does

File: mat64/doc.go [hunk 3]
```diff
@@ -2,38 +2,119 @@
...
+// constructs a *Dense from the result of a muliplication with any Matrix types,
+// not just *Dense. Where more restrictive requirements must be met, there are also the
+// Symmetric and Triangular interfaces. For example, in
+//  func (s *SymDense) AddSym(a, b Symmetric)
+// the Symmetric interface guarantees a symmetric result.
+//
+// Transposes
+//
+// The T method is used for transpostion. For example, c.Mul(a.T(), b) computes
+// c = a^T * b. The mat64 types implement this method using an implicit transpose.
```
Comment (vladimir-ch) [inline mat64/doc.go]: Perhaps elaborate what is meant by implicit transpose?

File: mat64/doc.go [hunk 4]
```diff
@@ -2,38 +2,119 @@
...
+// Matrix factorizations, such as the LU decomposition, typically have their own
+// specific data storage, and so are each implemented as a specific type. The
+// factorization can be computed through a call to Factorize
+//  var lu mat64.LU
+//  lu.Factorize(a)
+// The elements of the factorization can be extracted through methods on the
+// appropriate type, i.e. *TriDense.LFromLU and *TriDense.UFromLU. Alternatively,
+// they can be used directly, as in *Dense.SolveLU. Some factorizations can be
+// updated directly, without needing to update the original matrix and refactorize,
+// as in *LU.RankOne.
```
Comment (vladimir-ch) [inline mat64/doc.go]: Nice, I didn't know/missed that it existed.

File: mat64/doc.go [hunk 5]
```diff
@@ -2,38 +2,119 @@
...
+// calls will be partially executed in Go and partially executed in c.
+//
+// Type Switching
+//
+// The Matrix abstraction enables efficiency as well as interoperability. Go's
+// type reflection capabilities are used to choose the most efficient routine
+// given the specific concrete types. For example, in
+//  c.Mul(a, b)
+// if a and b both implement RawMatrixer, that is, they can be represented as a
+// blas64.General, blas64.Gemm (general matrix mulitplication) is called, while
```
Comment (vladimir-ch) [inline mat64/doc.go]: mulitplication = multiplication

File: mat64/doc.go [hunk 6]
```diff
@@ -2,38 +2,119 @@
...
+// change Matrix data, that matrix will be the receiver of a function.
+//
+// For convenience, a matrix may be used as both a receiver and as an input, i.e.
+//  a.Pow(a, 6)
+//  v.SolveVec(a.T(), v)
+// though in many cases this will cause an allocation. However, mat64 cannot
+// detect arbitrary overlap, and so care should be taken when using matrix views.
+// An exception to this rule is Copy, which does not allow a.Copy(a.T()).
+//
+// All functions may be called concurrently.
```
Comment (kortschak) [inline mat64/doc.go]: Is this necessary to say?

File: mat64/doc.go [hunk 7]
```diff
@@ -2,38 +2,119 @@
...
+//
+// Type Switching
+//
+// The Matrix abstraction enables efficiency as well as interoperability. Go's
+// type reflection capabilities are used to choose the most efficient routine
+// given the specific concrete types. For example, in
+//  c.Mul(a, b)
+// if a and b both implement RawMatrixer, that is, they can be represented as a
+// blas64.General, blas64.Gemm (general matrix mulitplication) is called, while
+// instead if b is a RawSymmetricer blas64.Symm is used (general-symmetric
```
Comment (vladimir-ch) [inline mat64/doc.go]: comma after RawSymmetricer?

File: mat64/doc.go [hunk 8]
```diff
@@ -2,38 +2,119 @@
...
+// abstract matrix type may be copied into a concrete type of the corresponding
+// value. If there are specific special cases that are needed, please submit a
+// pull-request or file an issue.
+//
+// Invariants
+//
+// Matrix input arguments to functions are never modified. If an operation would
+// change Matrix data, that matrix will be the receiver of a function.
+//
+// For convenience, a matrix may be used as both a receiver and as an input, i.e.
```
Comment (vladimir-ch) [inline mat64/doc.go]: It is good that this is mentioned, I would have been unsure myself otherwise.

File: mat64/doc.go [hunk 9]
```diff
@@ -2,38 +2,119 @@
...
+//  func (m *Dense) Mul(a, b Matrix)
+// constructs a *Dense from the result of a muliplication with any Matrix types,
+// not just *Dense. Where more restrictive requirements must be met, there are also the
+// Symmetric and Triangular interfaces. For example, in
+//  func (s *SymDense) AddSym(a, b Symmetric)
+// the Symmetric interface guarantees a symmetric result.
+//
+// Transposes
+//
+// The T method is used for transpostion. For example, c.Mul(a.T(), b) computes
```
Comment (vladimir-ch) [inline mat64/doc.go]: transposition

File: mat64/doc.go [hunk 10]
```diff
@@ -2,38 +2,119 @@
...
+// The Matrix Interface(s)
+//
+// The Matrix interface is the common link between the concrete types. The Matrix
+// interface is defined by three functions: Dims, which returns the dimensions
+// of the Matrix, At, which returns the element in the specified location, and
+// T, which will be discussed later. All of the concrete types can perform these
+// behaviors and so implement the interface. Methods and functions are designed
+// to use this interface, so in particular the method
+//  func (m *Dense) Mul(a, b Matrix)
+// constructs a *Dense from the result of a muliplication with any Matrix types,
```
Comment (vladimir-ch) [inline mat64/doc.go]: multiplication

File: mat64/doc.go [hunk 11]
```diff
@@ -2,38 +2,119 @@
...
+// they can be used directly, as in *Dense.SolveLU. Some factorizations can be
+// updated directly, without needing to update the original matrix and refactorize,
+// as in *LU.RankOne.
+//
+// BLAS and LAPACK
+//
+// BLAS and LAPACK are the standard APIs for linear algebra routines. Many
+// operations in mat64 are implemented using calls to the wrapper functions
+// in gonum/blas/blas64 and gonum/lapack/lapack64. By default, blas64 and
+// lapack64 call the native Go implementations of the routines.  Alternatively,
```
Comment (vladimir-ch) [inline mat64/doc.go]: Nitpicking: delete one space before Alternatively

File: mat64/doc.go (hunk 3 above)
Comment (btracey) [inline mat64/doc.go]: Is "See the Transpose type for more information" a good answer? I think this is one of those cases where the source code is easier to understand than the english.

File: mat64/doc.go (hunk 6 above)
Comment (btracey) [inline mat64/doc.go]: Probably not. In my mind, functions should always be written so they are concurrently callable, but I'm not sure what the general Go "standard wisdom" is.

File: mat64/doc.go (hunk 1 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 2 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 10 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 9 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 4 above)
Comment (btracey) [inline mat64/doc.go]: Yea, it's a nice function. The eventual goal was to work that into simplex...

File: mat64/doc.go (hunk 11 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 5 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 7 above)
Comment (btracey) [inline mat64/doc.go]: I don't think it needs one. It's not a clause.

File: mat64/doc.go (hunk 6 above)
Comment (btracey) [inline mat64/doc.go]: Deleted.

File: mat64/doc.go [hunk 12]
```diff
@@ -2,38 +2,117 @@
...
+// will be modified to have the correct size, allocating data if necessary.
+//  var c mat64.Dense // construct a new zero-sized matrix
+//  c.Mul(a, a)       // c is automatically adjusted to be 6×6
+//
+// The Matrix Interface(s)
+//
+// The Matrix interface is the common link between the concrete types. The Matrix
+// interface is defined by three functions: Dims, which returns the dimensions
+// of the Matrix, At, which returns the element in the specified location, and
+// T for returning a Transpos (discussed later). All of the concrete types can
```
Comment (kortschak) [inline mat64/doc.go]: Transpose

File: mat64/doc.go [hunk 13]
```diff
@@ -2,38 +2,117 @@
...
+// not just *Dense. Where more restrictive requirements must be met, there are also the
+// Symmetric and Triangular interfaces. For example, in
+//  func (s *SymDense) AddSym(a, b Symmetric)
+// the Symmetric interface guarantees a symmetric result.
+//
+// Transposes
+//
+// The T method is used for transposition. For example, c.Mul(a.T(), b) computes
+// c = a^T * b. The mat64 types implement this method using an implicit transpose,
+// see the Transpose type for more details. Note that some operations have a
```
Comment (kortschak) [inline mat64/doc.go]: Use a dash of some variety rather than a comma.

File: mat64/doc.go [hunk 14]
```diff
@@ -2,38 +2,117 @@
...
+// updated directly, without needing to update the original matrix and refactorize,
+// as in *LU.RankOne.
+//
+// BLAS and LAPACK
+//
+// BLAS and LAPACK are the standard APIs for linear algebra routines. Many
+// operations in mat64 are implemented using calls to the wrapper functions
+// in gonum/blas/blas64 and gonum/lapack/lapack64. By default, blas64 and
+// lapack64 call the native Go implementations of the routines. Alternatively,
+// it is possible to use c-based implementations of the APIs through the respective
```
Comment (kortschak) [inline mat64/doc.go]: upper case C.

File: mat64/doc.go [hunk 15]
```diff
@@ -2,38 +2,117 @@
...
+// cgo packages and "Use" routines. The Go implementation of LAPACK makes calls
```
Comment (kortschak) [inline mat64/doc.go]: s/routines/functions/

File: mat64/doc.go (hunk 6 above)
Comment (kortschak) [inline mat64/doc.go]: The general situation is that there is no shared global state unless explicitly stated, but values are not concurrency safe unless stated; functions are safe, values may not be. We fall squarely into that set.

File: mat64/doc.go (hunk 12 above)
Comment (btracey) [inline mat64/doc.go]: Argh. Sorry.

File: mat64/doc.go (hunk 13 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 14 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 15 above)
Comment (btracey) [inline mat64/doc.go]: Done.

File: mat64/doc.go (hunk 3 above)
Comment (vladimir-ch) [inline mat64/doc.go]: Ok

File: mat64/doc.go (hunk 7 above)
Comment (vladimir-ch) [inline mat64/doc.go]: It's not important, just that one line below there is a comma.

Comment (btracey) [inline mat64/doc.go]: Ah, I see. Removed below.

**Comment by kortschak** [comment]
This is a huge improvement over the original short doc, thank you.

LGTM

**Comment by vladimir-ch** [comment]
Same here, very useful comprehensive overview. Thanks!

**Pull request merged**